Bump Supported Python versions up to 3.12

marcharper opened the pull request:
Python 3.12 doesn't seem to be available yet in CI, will re-run tests later.

Update: some packages like distutils are no longer available in Python 3.12 so I had a make a few small changes. In one case we had a check for matplotlib versions below 2, but we're already requiring 3+, so there's no change in functionality.

marcharper committed:
Bump python version up to 3.12

drvinceknight commented:
Looks like this isn't passing because of some downstream libraries that aren't on 3.12 yet? I tried restarting the tests just now and it didn't make much difference.

marcharper committed:
Update load_data_.py to use pkutil

marcharper committed:
Remove distutils usage in plot.py since it's not available  in Python 3.12

marcharper committed:
Replace deprecated unittest method

drvinceknight committed:
Simplify the doctest.

The tests were failing on ubuntu
https://github.com/Axelrod-Python/Axelrod/actions/runs/8198844499/job/22423011180?pr=1428

This was caused by some rounding error.

I think this cleans things up.

drvinceknight commented:
@marcharper I've opened #1435 with a PR to your branch that I think cleans up the doctests.

marcharper commented:
`setuptools` is deprecated and removed in Python 3.12, so we have to either use a third party version or change how we do packaging :/

This link is helpful: https://gregoryszorc.com/blog/2023/10/30/my-user-experience-porting-off-setup.py/

IIUC we need to migrate from setup.py to pyproject.toml. But there are some choices that need to be made (e.g. build tool, setuptools is still an option).

Some options here: https://packaging.python.org/en/latest/guides/writing-pyproject-toml/

marcharper commented:
I pulled out the commits after the first into #1436 so we can merge those. I'll rebase the first commit onto a python-3.12 branch and we can figure out the packaging issues there.

marcharper commented:
Closing, let's discuss at #1437
